pkg/webhook/resources/virtualmachineimage/validator.go

```diff
@@ -70,8 +75,36 @@ func (v *virtualMachineImageValidator) Create(request *types.Request, newObj run
 
 	if newImage.Spec.SourceType == v1beta1.VirtualMachineImageSourceTypeDownload && newImage.Spec.URL == "" {
 		return werror.NewInvalidError(`url is required when image source type is "download"`, "spec.url")
-	} else if newImage.Spec.SourceType == v1beta1.VirtualMachineImageSourceTypeUpload && newImage.Spec.URL != "" {
-		return werror.NewInvalidError(`url should be empty when image source type is "upload"`, "spec.url")
+	} else if newImage.Spec.SourceType != v1beta1.VirtualMachineImageSourceTypeDownload && newImage.Spec.URL != "" {
+		return werror.NewInvalidError(`url should be empty when image source type is not "download"`, "spec.url")
+	} else if newImage.Spec.SourceType == v1beta1.VirtualMachineImageSourceTypeExportVolume {
```

Comment:
We can check required fields early. Otherwise, we may return "failed to check user permission" error when pvcName is missing which is not user-friendly.

Reply:
@gitlawr updated